File: 0238-product-of-array-except-self/0238-product-of-array-except-self.cs
public class Solution {
    public int[] ProductExceptSelf(int[] nums) {

        int n = nums.Length;
        int[] prefix = new int[n+1];
        int[] suffix = new int[n+1];
        int[] final = new int[n];

        prefix[0] = 1;
        suffix[n] = 1;

        for(int i=1; i<n+1; i++)
            prefix[i] = nums[i-1] * prefix[i-1];
        
        for(int i=n-1; i>0; i--)
            suffix[i] = nums[i] * suffix[i+1];

        for(int j=0; j<n; j++)
        {
            final[j] = prefix[j] * suffix[j+1];
        }

        return final;
    }
}

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        vector<int> pre(n+1);
        vector<int> suf(n+1);
        vector<int> res(n);
        pre[0] = 1;
        suf[n] = 1;
        
        for(int i=1; i<=n; i++) {
            pre[i] = nums[i-1] * pre[i-1];
        }
        
        for(int i=n-1; i>0; i--) {
            suf[i] = nums[i] * suf[i+1];
        }
        
        for(int i=0; i<n; i++) {
            res[i] = pre[i] * suf[i+1];
        }
        
        return res;
    }
};